Initial support for Ambiq Apollo4 Plus

msobkowski opened the pull request:
This PR provides initial support for the Ambiq Apollo4 Plus SoC and the Apollo4P EVB board.

The PR consists of:
- Adding Ambiq HAL module
- Ambiq Apollo4 Plus SoC support
- Ambiq Apollo4P EVB board support
- Ambiq System Timer (STIMER) driver
- Changes to PL011 UART driver to support variant present in Apollo4 Plus SoC

Initial support was verified on hardware by running the following samples:
- hello_world
- philosophers
- subsys/shell/shell_module

This PR should be merged only after https://github.com/zephyrproject-rtos/hal_ambiq/pull/1 is merged and hal_ambiq revision in west.yml is bumped to the newest one.

msobkowski committed:
soc: arm: ambiq: apollo4: Add support for Apollo4 Plus SoC

Add all required parts (new SoC family/series, device tree) for
the Ambiq Apollo4 Plus SoC.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
boards: Add support for the Ambiq Apollo4P EVB board

This commit adds support for the Ambiq Apollo4P EVB board (AMAP4PEVB).

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
west.yml: add Ambiq HAL module

Add the Ambiq HAL module to the West manifest.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
modules: add Kconfig for Ambiq HAL module

This commit introduces the Ambiq HAL module required for the
Ambiq Apollo4 Plus SoC support.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
drivers: timer: Add driver for Ambiq system timer (STIMER)

This commit addst support for the system timer peripheral which
can be found in Apollo4 SoCs.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
soc: arm: ambiq: apollo4x: configure STIMER

Disable SysTick when STIMER is in use and configure
SYS_CLOCK_HW_CYCLES_PER_SEC setting for it.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
MAINTAINERS: Add entry for Ambiq platforms

Add people responsible for maintaining Zephyr on Ambiq SoCs.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

gmarull commented on soc/arm/ambiq/apollo4x/soc.c:
```diff
@@ -0,0 +1,70 @@
+
+LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
+
+am_hal_gpio_pincfg_t pin_rx_config = {
+	.GP.cfg_b.uFuncSel             = AM_HAL_PIN_47_UART0RX,
+	.GP.cfg_b.eGPInput             = AM_HAL_GPIO_PIN_INPUT_NONE,
+	.GP.cfg_b.eGPRdZero            = AM_HAL_GPIO_PIN_RDZERO_READPIN,
+	.GP.cfg_b.eIntDir              = AM_HAL_GPIO_PIN_INTDIR_NONE,
+	.GP.cfg_b.eGPOutCfg            = AM_HAL_GPIO_PIN_OUTCFG_DISABLE,
+	.GP.cfg_b.eDriveStrength       = AM_HAL_GPIO_PIN_DRIVESTRENGTH_0P1X,
+	.GP.cfg_b.uSlewRate            = 0,
+	.GP.cfg_b.ePullup              = AM_HAL_GPIO_PIN_PULLUP_NONE,
+	.GP.cfg_b.uNCE                 = 0,
+	.GP.cfg_b.eCEpol               = AM_HAL_GPIO_PIN_CEPOL_ACTIVELOW,
+	.GP.cfg_b.uRsvd_0              = 0,
+	.GP.cfg_b.ePowerSw             = AM_HAL_GPIO_PIN_POWERSW_NONE,
+	.GP.cfg_b.eForceInputEn        = AM_HAL_GPIO_PIN_FORCEEN_NONE,
+	.GP.cfg_b.eForceOutputEn       = AM_HAL_GPIO_PIN_FORCEEN_NONE,
+	.GP.cfg_b.uRsvd_1              = 0,
+};
```
please implement a proper pinctrl driver

msobkowski replied:
Thanks for a quick response, will create a proper pinctrl driver.

msobkowski replied:
@gmarull: I have updated the PR to include a pinctrl driver

msobkowski committed:
drivers: pinctrl: Add pinctrl driver for Apollo4

This commit addst pinctrl support for Apollo4 SoCs.

Co-authored-by: Mateusz Sierszulski <msierszulski@antmicro.com>
Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
dts: ambiq: apollo4p: instantiate pinctrl

This commit instantiates pinctrl node in the dts file for Apollo4
Plus SoC.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
dts: vendor-prefixes: Add Ambiq Micro vendor prefix

Added Ambiq Micro vendor prefix to enable dts bindings for peripherals in
Ambiq SoCs.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski commented:
I've added a commit (f9012b4f7b744cc6b14a55edf3c64d3e91308a94) that sets remote for Ambiq HAL in west.yml to github.com/antmicro (where the repo is available currently) to enable building Zephyr for Apollo4P. Request for HAL module repo is tracked here: #59473.

gmarull commented on dts/arm/ambiq/ambiq_apollo4p.dtsi:
```diff
@@ -25,6 +25,10 @@
 		reg = <0x10000000 0x60000>;
 	};
+
+	pinctrl: pin-controller {
```
why is this a _virtual device_? if not device specific (as Nordic), it should have a register

msobkowski replied:
Added base address to the pinctrl node.

gmarull commented on boards/arm/apollo4p_evb/apollo4p_evb_defconfig:
```diff
@@ -3,3 +3,4 @@
 # Copyright (c) 2023 Antmicro <www.antmicro.com>
 
 CONFIG_SOC_SERIES_APOLLO4X=y
+CONFIG_PINCTRL=y
```
this is not required (some boards still do this for historical reasons), drivers need to `select PINCTRL` instead.

msobkowski replied:
Removed `CONFIG_PINCTRL=y` from defconfig, [added `select PINCTRL` to the UART driver](https://github.com/zephyrproject-rtos/zephyr/blob/4aace128e0d88e660dcdc4a4473f8736927196d0/drivers/serial/Kconfig.pl011#L10) instead.

gmarull commented on drivers/serial/uart_pl011.c:
```diff
@@ -125,6 +125,10 @@ struct pl011_data {
 #define PL011_CR_UARTEN		BIT(0)	/* enable uart operations */
 #define PL011_CR_SIREN		BIT(1)	/* enable IrDA SIR */
 #define PL011_CR_SIRLP		BIT(2)	/* IrDA SIR low power mode */
+#ifdef CONFIG_UART_PL011_AMBIQ_CLK_CR_REG
+#define PL011_CR_CLKEN		BIT(3)	/* clock enable */
+#define PL011_CR_CLKSEL		GENMASK(6, 4) /* clock select */
+#endif
```
I'm not the UART maintainer, but IMO vendor quirks on generic IP drivers should be split in a separate file.

msobkowski replied:
I've added support for vendor quirks in the PL011 driver, and moved Ambiq-specific quirks to a separate file.

gmarull commented on drivers/serial/uart_pl011.c:
```diff
@@ -432,6 +480,15 @@ static int pl011_init(const struct device *dev)
+		 */
+#ifdef CONFIG_UART_PL011_AMBIQ_CLK_CR_REG
+		pl011_enable_clk(dev);
+		pl011_clk_set(dev, config->sys_clk_freq);
```
error handling missing here

msobkowski replied:
Return value of the clk_set function is handled now (the function has a different name now after implementing vendor quirks): https://github.com/zephyrproject-rtos/zephyr/blob/4aace128e0d88e660dcdc4a4473f8736927196d0/drivers/serial/uart_pl011_ambiq.h#L51

gmarull commented on drivers/serial/Kconfig.pl011:
```diff
@@ -21,4 +21,11 @@ config UART_PL011_SBSA
+	bool "Clock configuration via the CR register"
+	default y if SOC_APOLLO4P
+	help
+	  Enable configuring clock via the CR register. This feature
+	  is present in Ambiq Apollo4 SoCs.
```
is this an Ambiq specific modification to the PL011 IP?

msobkowski replied:
Yes, this modification to PL011 core is specific to Ambiq SoCs.

gmarull commented on boards/arm/apollo4p_evb/Kconfig.board:
```diff
@@ -0,0 +1,7 @@
+#
+# Copyright (c) 2023 Antmicro <www.antmicro.com>
+
+config BOARD_APOLLO4P_EVB
```
note: board in the first commit is useless, add it later in the sequence (e.g. when uart can print hello world)

msobkowski replied:
Reordered the commits so that the board is added after drivers are in place.

gmarull commented on boards/arm/apollo4p_evb/apollo4p_evb.dts:
```diff
@@ -0,0 +1,14 @@
+	model = "Ambiq Apollo4 Plus evaluation board";
+	compatible = "ambiq,apollo4p_evb";
+	chosen {
+		zephyr,sram = &sram0;
+		zephyr,flash = &flash0;
+		zephyr,console = &uart0;
+		zephyr,shell-uart = &uart0;
+		zephyr,uart-pipe = &uart0;
```
related to my previous commit, where's uart0?

msobkowski replied:
After commits reordering `uart0` is now present when the board is added.

gmarull commented on dts/bindings/pinctrl/ambiq,apollo4-pinctrl.yaml:
```diff
@@ -0,0 +1,92 @@
+          - "0.75"
+          - "1.0"
+        default: "0.5"
+        description: |
+          The drive strength of a pin, relative to full-driver strength.
+          The default value is 0.5.
+      slew-rate:
+        default: "slow"
+        type: string
+        enum:
+          - "slow"
+          - "fast"
+        description: |
+          Select slew rate for pin.
```
defaults miss justification here

msobkowski replied:
Updated default values and added justification (default values are now the same as reset values).

gmarull commented on dts/arm/ambiq/ambiq_apollo4p.dtsi:
```diff
@@ -26,6 +26,50 @@
+			compatible = "arm,pl011";
+			reg = <0x4001c000 0x1000>;
+			interrupts = <15 0>;
+			interrupt-names = "irq_15";
```
question: is this the real IRQ name?

msobkowski replied:
Updated IRQ names to match IRQ descriptions located in the "ARM Cortex-M4 Vector Table for Apollo4 Plus SoC" table from the Apollo4 Plus Programmer’s Guide.

gmarull commented on soc/arm/ambiq/apollo4x/soc.c:
```diff
@@ -16,7 +17,15 @@
+
+#define DT_DRV_COMPAT arm_pl011
+	DT_INST_FOREACH_STATUS_OKAY(AMBIQ_UART_PWRCTRL_ENABLE)
+#undef DT_DRV_COMPAT
```
-1 on this, please find a better solution (e.g. driver quirk)

msobkowski replied:
Moved enabling power for UART to a driver quirk: https://github.com/zephyrproject-rtos/zephyr/blob/4aace128e0d88e660dcdc4a4473f8736927196d0/drivers/serial/uart_pl011_ambiq.h#L54-L68

gmarull commented on soc/arm/ambiq/apollo4x/soc.c:
```diff
@@ -16,7 +17,15 @@
 
 LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
 
+#define AMBIQ_UART_PWRCTRL_ID(uart) AM_HAL_PWRCTRL_PERIPH_UART ## uart
+#define AMBIQ_UART_PWRCTRL_ENABLE(idx) \
+	am_hal_pwrctrl_periph_enable(AMBIQ_UART_PWRCTRL_ID(idx));
```
DT instances order is not guaranteed, ie, inst 0 may not be uart0

msobkowski replied:
Added a dt property in the Ambiq-specific UART binding that denotes the instance number: https://github.com/zephyrproject-rtos/zephyr/blob/4aace128e0d88e660dcdc4a4473f8736927196d0/dts/bindings/serial/ambiq%2Cuart.yaml#L17-L20

gmarull commented on soc/arm/ambiq/apollo4x/soc.c:
```diff
@@ -4,6 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include "hal/am_hal_utils.h"
```
not a local header, please don't use `""`

msobkowski replied:
Removed this include entirely, it is not needed

gmarull commented on drivers/timer/ambiq_stimer.c:
```diff
@@ -0,0 +1,153 @@
+#include <zephyr/logging/log.h>
+
+#include <am_mcu_apollo.h>
+#include <am_bsp.h>
```
"bsp" includes?

msobkowski replied:
Removed all BSP includes.

gmarull commented on soc/arm/ambiq/Kconfig.defconfig:
```diff
@@ -3,3 +3,9 @@
 
 source "soc/arm/ambiq/*/Kconfig.defconfig.series"
+
+config CORTEX_M_SYSTICK
+	default n if AMBIQ_STIMER_TIMER
```
is it default y?

msobkowski replied:
It is now - I've added `select CPU_CORTEX_M_HAS_SYSTICK`, to Kconfig.series (which enables `CONFIG_CORTEX_M_SYSTICK`): https://github.com/zephyrproject-rtos/zephyr/blob/a0858dcc312eb6115f9f7e713c6d7688524fc5bd/soc/arm/ambiq/apollo4x/Kconfig.series#L12

gmarull replied:
this is indicating SoC has systick, but not enabling it

msierszulski replied:
i removed `select CPU_CORTEX_M_HAS_SYSTICK`.

gmarull commented on soc/arm/ambiq/apollo4x/Kconfig.series:
```diff
@@ -0,0 +1,17 @@
+	select CPU_HAS_FPU
+	select SOC_FAMILY_AMBIQ
+	select HAS_SWO
+	select AMBIQ_HAL
```
this needs to be selected by drivers if needed. see e.g. st/gd

msobkowski replied:
I've modified HAL Kconfig so that specific HAL drivers can be selected independently by Zephyr drivers:
https://github.com/zephyrproject-rtos/zephyr/blob/4aace128e0d88e660dcdc4a4473f8736927196d0/drivers/timer/Kconfig.ambiq#L11
https://github.com/zephyrproject-rtos/zephyr/blob/4aace128e0d88e660dcdc4a4473f8736927196d0/drivers/pinctrl/Kconfig.ambiq#L9

But the SoC init also calls Ambiq HAL functions, so I think the main HAL config needs to be selected here?

gmarull commented on soc/arm/ambiq/apollo4x/soc.c:
```diff
@@ -0,0 +1,22 @@
+
+void z_arm_platform_init(void)
+{
+	am_bsp_low_power_init();
```
`bsp`? what does this function do? we do not allow "bsp-like" code in Zephyr (we have `boards/`)

msobkowski replied:
I've replaced the BSP function with calls to actual HAL functions:
https://github.com/zephyrproject-rtos/zephyr/blob/4aace128e0d88e660dcdc4a4473f8736927196d0/soc/arm/ambiq/apollo4x/soc.c#L19-L21

gmarull commented on soc/arm/ambiq/apollo4x/soc.h:
```diff
@@ -0,0 +1,14 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __SOC_H__
+
+#include <am_mcu_apollo.h>
+#include <am_bsp.h>
+
+#define __SOC_H__
```
wrong header guard. why is bsp needed here?

msobkowski replied:
Fixed the header guard. BSP include is not needed here, I've removed it.

msobkowski committed:
dts: arm: ambiq: apollo4p: instantiate UARTs

This commit adds PL011 UART instances to the apollo4p dts.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
drivers: serial: pl011: move register definitions into a header

The registers definitions will be needed when adding vendor-specific quirks
to this driver, so this commits moves them to a dedicated header file.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
dts: bindings: Add new binding ambiq,uart

UART controllers present on Ambiq SoCs implement a PL011 compatible
interface, with some minor differences that require certain quirks.
To support them a dedicated compatible is needed.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
drivers: serial: pl011: add definitions for CLKEN/CLKSEL registers

Add definitions for CLKEN/CLKSEL registers, which are used to control
peripheral clock on the variant of the PL011 UART present in Ambiq SoCs.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski committed:
drivers: serial: pl011: Add support for Ambiq UART

UART controller present in Ambiq SoCs is mostly compatible with PL011, but
requires some quirks that are implemented in this commit:
- the peripheral needs to be powered on first, via the PWRCTRL core,
- peripheral clock needs to be enabled and configured via the CLKEN/CLKSEL.
  registers.

The quirks mechanism was inspired by support for STM32F4 SoC in the
usb_dc_dw driver (fce0b85ecaed71f43c018391909a23cefb747ac6).

Co-authored-by: Mateusz Sierszulski <msierszulski@antmicro.com>
Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>
Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

msobkowski committed:
drivers: serial: pl011: Select PINCTRL for Apollo4 SoC family

PINCTRL needs to be enabled for the driver to work on the Apollo4 family
SoCs.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

gmarull commented on boards/arm/apollo4p_evb/apollo4p_evb_defconfig:
```diff
@@ -0,0 +1,10 @@
+CONFIG_UART_CONSOLE=y
+CONFIG_SERIAL=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_HAS_AMBIQ_HAL=y
```
why is board selecting HAL?

msierszulski replied:
I changed it so that this config is selected in SoC.

gmarull replied:
an SoC doesn't have a HAL ;-)

msierszulski replied:
Can you explain what is the intended way to select this config? We use the function from the hal in [soc.c](https://github.com/zephyrproject-rtos/zephyr/pull/59470/files#diff-52af905f6fb03f5163df72c5aaae927e0e5bb3e2ad3f7e3d10f8f184cd23efdfR18-R20). In addition, each driver selects a different symbol provided for a specific driver, such as [stimer](https://github.com/zephyrproject-rtos/zephyr/pull/59470/files#diff-f72da9c776aca3dab9ea5c6b6659ce0101011ba230460023e0c1b3c27219a2edR11).

gmarull commented on dts/bindings/serial/ambiq,uart.yaml:
```diff
@@ -0,0 +1,20 @@
+  pinctrl-names:
+    required: true
+
+  ambiq,uart-instance:
+    type: int
+    description: |
+      Instance number of the UART peripheral
```
this has nothing to do with hw description, please remove.

msierszulski replied:
In fact, it has to do with the description of hw, they are not just logical instances. 
Using instances from Zephyr for example. `DT_INST_FOREACH_STATUS_OKAY` we will get an instance of 0 for `uart2` if only `uart2` has `status = "okay"`. By using `ambiq,uart-instance` we are sure that we are passing the correct hardware instance to the driver.

msierszulski replied:
I think we misunderstood each other. The `ambiq,uart-instance` field is specific to each uart instance in DTS. Adding this field ensures that it is properly handled by the driver regardless of the driver instance.

gmarull replied:
Again, DT is reserved for hardware description, not driver-specific needs. Instance numbers are a common sign of halisms in Devicetree because HALs tend to rely on those numbers.

I haven't checked in detail, but this kind of functionality is usually handled by reset/clock controllers where you have in DT a cell where to encode information (e.g. the peripheral bit in enablement register).

msierszulski replied:
Ok, I get it. We'll get rid of this instance.

msierszulski replied:
Removed `ambiq,uart-instance`.

gmarull reviewed (commented):
note: HAL contains unnecessary pieces like https://github.com/antmicro/hal_ambiq/tree/main/mcu/apollo4p/hal/mcu/iar or https://github.com/antmicro/hal_ambiq/tree/main/devices, these should be dropped to minimize module size (to avoid continued growing of modules with unnecessary stuff)

msierszulski commented:
> note: HAL contains unnecessary pieces like https://github.com/antmicro/hal_ambiq/tree/main/mcu/apollo4p/hal/mcu/iar or https://github.com/antmicro/hal_ambiq/tree/main/devices, these should be dropped to minimize module size (to avoid continued growing of modules with unnecessary stuff)

Unnecessary code has been removed.

gmarull commented on drivers/serial/uart_pl011_ambiq.h:
```diff
@@ -0,0 +1,72 @@
+	return pl011_ambiq_clk_set(dev, clk);
+}
+
+#define AMBIQ_UART_PWRCTRL_ENABLE(idx)                                                             \
+	am_hal_pwrctrl_periph_enable(AM_HAL_PWRCTRL_PERIPH_UART0 + idx);
```
instance number 0 doesn't necessarily mean UART0

I think this can be handled better with something like this: https://elixir.bootlin.com/linux/latest/source/arch/arm/boot/dts/stm32h743.dtsi#L531

msierszulski replied:
I changed the handling of pwrctrl as you stated.

gmarull commented on drivers/serial/uart_pl011_ambiq.h:
```diff
@@ -0,0 +1,72 @@
+	am_hal_pwrctrl_periph_enable(AM_HAL_PWRCTRL_PERIPH_UART0 + idx);
+
+#define QUIRK_AMBIQ_UART_DEFINE(n)                                                                 \
+	static int pwr_on_ambiq_uart_##n(void)                                                     \
+	{                                                                                          \
+		for (int i = 0; i < AM_HAL_PWRCTRL_PERIPH_MAX; i++) {                              \
+			if ((uint32_t)UARTn(i) == DT_INST_REG_ADDR(n)) {                           \
+				return AMBIQ_UART_PWRCTRL_ENABLE(i);                               \
+			}                                                                          \
+		}                                                                                  \
+		return -ENOTSUP;                                                                   \
+	}
```
see my previous comments, I think there are better options than this hack

msierszulski replied:
I changed the handling of pwrctrl as you stated.

msierszulski replied:
I changed the handling of pwrctrl as you stated.

msierszulski committed:
dts: bindings: Add new ambiq,pwrctrl binding

This commit adds Ambiq power control dedicated compatible.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

gmarull commented on dts/bindings/pinctrl/ambiq,apollo4-pinctrl.yaml:
```diff
@@ -0,0 +1,107 @@
+        default: "slow"
+        description: |
+          Select slew rate for a pin. The default is slow, which is the reset value.
+      bias-value:
+        type: string
+        enum:
+          - "0"
+          - "pulldown_50k"
+          - "pullup_1_5k"
+          - "pullup_6k"
+          - "pullup_12k"
+          - "pullup_12k"
+          - "pullup_24k"
+          - "pullup_50k"
+          - "pullup_100k"
+        default: "0"
+        description: |
+          The pullup/pulldown resistor value. The default value is reset value.
```
pull-up/down enablement should still follow `bias-pull-up/down`. Value may be selected using an extra property, e.g. `ambiq,pull-up-ohms` and value be in ohms (you can still use an enumerated integer to rely on DT enum macros if preferred)

msierszulski replied:
Added `bias-pull-up/down`. Also added `ambiq,pull-up-ohms` for pull-up resitor value configuration.

gmarull commented on dts/bindings/serial/ambiq,uart.yaml:
```diff
@@ -0,0 +1,20 @@
+  pinctrl-names:
+    required: true
+
+  pwrcfg:
+    type: phandle-array
+    specifier-space: pwrcfg
+    description: Power configuration
```
note: you can likely have a common file defining this, so all ambiq IP peripherals can re-use it

msierszulski replied:
Sure. I moved it to `ambiq-pwrcfg.yaml`

gmarull commented on dts/arm/ambiq/ambiq_apollo4p.dtsi:
```diff
@@ -0,0 +1,98 @@
+			interrupt-names = "UART0";
+			status = "disabled";
+			clocks = <&uartclk>;
+			pwrcfg = <&pwrcfg 0x4 0x200>;
```
`ambiq,pwrcfg`

msierszulski replied:
Done.

gmarull commented on modules/Kconfig.ambiq:
```diff
@@ -0,0 +1,19 @@
+	depends on SOC_SERIES_APOLLO4X
+	help
+	  Use the GPIO driver from Ambiq HAL
+
```
`#if HAS_AMBIQ_HAL`

msierszulski replied:
Added `IF` for other configs.

gmarull commented on modules/Kconfig.ambiq:
```diff
@@ -0,0 +1,19 @@
+#
+# Copyright (c) 2023 Antmicro <www.antmicro.com>
+
+config HAS_AMBIQ_HAL
```
`HAS_` is really reserved to indicate certain features are available (e.g. `HAS_FPU`), even though this has been abused by some HALs. Selectable symbols should also be promptless.

What I'd suggest here is to:

1. rename to `AMBIQ_HAL` (not promptless, as `HAS_` should be)
2. Drivers (dt-based, so can use select) do: `select AMBIQ_HAL` and `select AMBIQ_HAL_USE_XXX`.

msierszulski replied:
Ok, now I get it. I changed that also.

gmarull commented on soc/arm/ambiq/apollo4x/Kconfig.soc:
```diff
@@ -0,0 +1,13 @@
+
+config SOC_APOLLO4P
+	bool "Apollo4P"
+	select HAS_AMBIQ_HAL
```
already selected by soc series?

msierszulski replied:
Yes, removed it.

gmarull commented on soc/arm/ambiq/apollo4x/soc.c:
```diff
@@ -0,0 +1,21 @@
+
+LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
+
+void z_arm_platform_init(void)
+{
+	am_hal_pwrctrl_low_power_init();
+	am_hal_pwrctrl_control(AM_HAL_PWRCTRL_CONTROL_SIMOBUCK_INIT, 0);
+	am_hal_rtc_osc_disable();
+}
```
question: is there any reason to not use a SYS_INIT as many other SoCs do?

msierszulski replied:
Added `SYS_INIT`.

gmarull commented on soc/arm/ambiq/apollo4x/soc.c:
```diff
@@ -0,0 +1,21 @@
+#include <zephyr/arch/cpu.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <soc.h>
+
+LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
```
this is not logging

msierszulski replied:
I removed that.

gmarull commented on soc/arm/ambiq/apollo4x/soc.c:
```diff
@@ -0,0 +1,21 @@
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <soc.h>
```
soc.h is a header that exists purely because of how CMSIS is included by the ARM port, you should include what you need, ie, hal headers.

msierszulski replied:
Removed this include and added hal header.

gmarull commented on soc/arm/ambiq/apollo4x/soc.c:
```diff
@@ -0,0 +1,21 @@
+ */
+
+#include <zephyr/arch/cpu.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
```
not used

msierszulski replied:
Removed.

msierszulski committed:
dts: bindings: Add new ambiq-pwrcfg binding

This commit adds Ambiq power configuration dedicated
compatible.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

gmarull commented on dts/arm/ambiq/ambiq_apollo4p.dtsi:
```diff
@@ -0,0 +1,32 @@
+		};
+	};
+
+	// MRAM region
+	flash0: flash@18000 {
+		compatible = "soc-nv-flash";
+		reg = <0x00018000 0x1e8000>;
+	};
+
+	// TCM + SRAM region
```
nit: no `//` comments

msierszulski replied:
Changed the comments.

gmarull commented on soc/arm/ambiq/apollo4x/soc.c:
```diff
@@ -0,0 +1,21 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/arch/cpu.h>
```
not needed

msierszulski replied:
Deleted.

gmarull commented on soc/arm/ambiq/apollo4x/soc.c:
```diff
@@ -0,0 +1,21 @@
+static int arm_apollo4_init(void)
+{
+	am_hal_pwrctrl_low_power_init();
+	am_hal_pwrctrl_control(AM_HAL_PWRCTRL_CONTROL_SIMOBUCK_INIT, 0);
```
if controlling a buck converter, this likely needs a regulator driver. could you provide some reference/guiding here?

msierszulski replied:
It turned out that the default configuration is sufficient for basic operations. We'll contribute a proper regulator driver in a separate PR.

gmarull commented on drivers/pinctrl/Kconfig.ambiq:
```diff
@@ -0,0 +1,12 @@
+	bool "Ambiq Apollo4 pin controller driver"
+	default y
+	depends on DT_HAS_AMBIQ_APOLLO4_PINCTRL_ENABLED
+	depends on AMBIQ_HAL
```
`select`?

msierszulski replied:
I changed it.

gmarull commented on dts/bindings/serial/ambiq,uart.yaml:
```diff
@@ -0,0 +1,15 @@
+    required: true
+
+  pinctrl-names:
+    required: true
```
maybe require the ambiq,pwrctrl property as well?

msierszulski replied:
This is a good idea. Thanks!

gmarull commented on drivers/serial/Kconfig.pl011:
```diff
@@ -7,6 +7,7 @@ menuconfig UART_PL011
 	depends on DT_HAS_ARM_PL011_ENABLED || DT_HAS_ARM_SBSA_UART_ENABLED
 	select SERIAL_HAS_DRIVER
 	select SERIAL_SUPPORT_INTERRUPT
+	select PINCTRL if SOC_SERIES_APOLLO4X
```
`select PINCTRL if DT_HAS_AMBIQ_UART_ENABLED`

gmarull commented on dts/arm/ambiq/ambiq_apollo4p.dtsi:
```diff
@@ -26,7 +26,59 @@
+	};
+
 	soc {
+		uart0: uart@4001c000 {/* UART0 */
```
nit: redundant trailing comment

msierszulski replied:
I deleted those comments.

gmarull commented on dts/arm/ambiq/ambiq_apollo4p.dtsi:
```diff
@@ -26,7 +26,59 @@
 		reg = <0x10000000 0x2C0000>;
 	};
 
+	uartclk: apb-pclk {
+		compatible = "fixed-clock";
+		clock-frequency = <24000000>;
+		#clock-cells = <0>;
+	};
```
nit: create `clocks` node?

msierszulski replied:
Added `clocks` node.

gmarull commented on drivers/timer/Kconfig.ambiq:
```diff
@@ -0,0 +1,16 @@
+	bool "Ambiq STIMER system clock driver"
+	default y
+	depends on DT_HAS_AMBIQ_STIMER_ENABLED
+	depends on AMBIQ_HAL
```
`select`?

msierszulski replied:
Changed.

gmarull commented on drivers/timer/Kconfig.ambiq:
```diff
@@ -0,0 +1,16 @@
+	help
+	  If you enable this, STIMER will be used to provide hw_cycles and
+	  kernel ticks instead of Cortex-M SysTick.
+
+	  If unsure, say 'N'.
```
looks redundant (option dependent on dt selection)

gmarull commented on drivers/timer/ambiq_stimer.c:
```diff
@@ -0,0 +1,153 @@
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <soc.h>
```
soc.h not needed

msierszulski replied:
Deleted.

gmarull commented on drivers/timer/ambiq_stimer.c:
```diff
@@ -0,0 +1,153 @@
+ *
+ */
+
+#include <zephyr/device.h>
```
device API not used here

msierszulski replied:
Changed for `#include <zephyr/init.h>`.

gmarull commented on drivers/timer/ambiq_stimer.c:
```diff
@@ -0,0 +1,153 @@
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <soc.h>
+#include <zephyr/drivers/clock_control.h>
```
clock_control API not used here

msierszulski replied:
Deleted.

gmarull commented on drivers/timer/ambiq_stimer.c:
```diff
@@ -0,0 +1,153 @@
+#include <zephyr/sys_clock.h>
+#include <zephyr/irq.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/logging/log.h>
```
logging not used

msierszulski replied:
Deleted.

gmarull reviewed (commented):
lgtm, just a few minor comments

gmarull commented on dts/arm/ambiq/ambiq_apollo4p.dtsi:
```diff
@@ -0,0 +1,100 @@
+		reg = <0x10000000 0x2C0000>;
+	};
+
+	pwrcfg: pwrcfg@40021000 {
+		compatible = "ambiq,pwrctrl";
+		reg = <0x40021000 0x400>;
+		#pwrcfg-cells = <2>;
+	};
+
+	soc {
+		stimer0: stimer@40008800 {
```
shouldn't this one be under soc?

msierszulski replied:
It should be. I moved it.

gmarull reviewed (dismissed):
lgtm, thanks for the continued effort

note: skipped reviewing details for timer/uart drivers

carlescufi commented:
@msobkowski please take a look at the comment in https://github.com/zephyrproject-rtos/hal_ambiq/pull/1

msierszulski commented:
Thanks @carlescufi, I replied to this comment.

msierszulski commented:
Hi @gmarull @carlescufi, can you take another look at this PR? Thanks.

msierszulski commented:
It looks like the `DNM` label is added automatically because the `hal_ambiq` is currently not present in the west.yml file on the main branch, and this triggers Zephyrbot to add the `DNM` label: https://github.com/zephyrproject-rtos/zephyr/actions/runs/5727067842/job/15518735602?pr=59470#step:3:102. It seems that this is handled by the [manifest](https://github.com/zephyrproject-rtos/action-manifest) action.

Normally, when the HAL PR is merged, i.e. `west.yml` doesn't point to the `pull/x/head` revision, the DNM label is dropped automatically, so I have removed it manually in this case.

kgugala reviewed (approved):
LGTM
